File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    graph = [[] for _ in range(n+1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def bfs(node):
        cnt = 1
        q = deque()
        q.append(node)
        visited = [0] * (n + 1)
        visited[node] = 1
        
        while q:
            now = q.popleft()
            if len(graph[now]) > 0:
                for x in graph[now]:
                    if not visited[x]:
                        visited[x] = 1
                        q.append(x)
                        cnt += 1
        return cnt
    
    answer = n
    for a, b in wires:
        graph[a].remove(b)
        graph[b].remove(a)

        answer = min(answer, abs(bfs(a) - bfs(b)))
        
        graph[a].append(b)
        graph[b].append(a)
        
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    public static ArrayList<Integer>[] graph;
    public static int N;
    
    public int solution(int n, int[][] wires) {
        N = n + 1;
        graph = new ArrayList[N];
        for(int i = 1; i < N; i++){
            graph[i] = new ArrayList<>();
        }
        
        for(int[] w : wires){
            graph[w[0]].add(w[1]);
            graph[w[1]].add(w[0]);
        }
        
        int answer = Integer.MAX_VALUE;
        for(int[] w : wires){
            graph[w[0]].remove(Integer.valueOf(w[1]));
            graph[w[1]].remove(Integer.valueOf(w[0]));
            
            answer = Math.min(answer, Math.abs(bfs(w[0]) - bfs(w[1])));
                              
            graph[w[0]].add(w[1]);
            graph[w[1]].add(w[0]);
        }
        return answer;
    }
    
    public int bfs(int start){
        boolean[] visited = new boolean[N];
        Queue<Integer> q = new LinkedList<>();
        visited[start] = true;
        q.add(start);
        
        int cnt = 1;
        while(!q.isEmpty()){
            int now = q.poll();
            for(int next : graph[now]){
                if(!visited[next]){
                    visited[next] = true;
                    q.add(next);
                    cnt++;
                }
            }
        }
        return cnt;
    }
}